File: Natixis Craftsmanship Practices/Breaking Dependencies/BankTransfer/cpp/src/MegaTransferEngine.cpp
#include "MegaTransferEngine.h"

#include <memory>
#include <stdexcept>

#include "HttpRiskClient.h"
#include "NativeSql.h"

bool MegaTransferEngine::doIt(const TData &d, const std::string &channel) {
    if (GlobalState::maintenance) {
        throw std::logic_error("Maintenance");
    }

    GlobalState::transferCount++;

    int fee = 0;

    if (channel == "MOBILE") {
        fee += 2;
    } else if (channel == "WEB") {
        fee += 1;
    }

    if (d.d) {
        fee--;
    }

    int net = d.c - fee;
    auto sql = std::make_shared<NativeSql>();

    int balance;
    if (GlobalState::cache[d.a]) {
        balance = GlobalState::cache[d.a];
    } else {
        balance = sql->queryBalance(d.a);
        GlobalState::cache[d.a] = balance;
    }

    auto http = std::make_shared<HttpRiskClient>();

    if (http->risky(d.a, net)) {
        if (!d.d) {
            return false;
        }
    }

    sql->updateBalance(d.a, balance - d.c);

    if (net % 2 == 0) {
        sql->updateBalance(d.b, net);
    } else {
        sql->updateBalance(d.b, net - 1);
        sql->updateBalance(d.b, 1);
    }

    if (GlobalState::transferCount > 100) {
        GlobalState::cache.clear();
    }

    return true;
}

File: Natixis Craftsmanship Practices/Breaking Dependencies/01 BankTransfer/csharp/BankTransfer/MegaTransferEngine.cs
using BankTransfer.External;

namespace BankTransfer;

public class MegaTransferEngine
{
    public bool DoIt(TData d, string channel)
    {
        if (GlobalState.Maintenance) throw new IllegalStateException("Maintenance");

        GlobalState.TransferCount++;

        var fee = 0;

        if ("MOBILE".Equals(channel))
        {
            fee += 2;
        }
        else
        {
            if ("WEB".Equals(channel)) fee += 1;
        }

        if (d.D) fee--;

        var net = d.C - fee;

        // BUG volontaire : net peut être négatif mais on continue
        var sql = new NativeSql();

        int balance;
        if (GlobalState.Cache.ContainsKey(d.A))
        {
            balance = GlobalState.Cache[d.A];
        }
        else
        {
            balance = sql.QueryBalance(d.A);
            GlobalState.Cache[d.A] = balance;
        }

        var http = new HttpRiskClient();

        if (http.Risky(d.A, net))
            if (!d.D)
                return false;

        sql.UpdateBalance(d.A, balance - d.C);

        if (net % 2 == 0)
        {
            sql.UpdateBalance(d.B, net);
        }
        else
        {
            sql.UpdateBalance(d.B, net - 1);
            sql.UpdateBalance(d.B, 1);
        }

        if (GlobalState.TransferCount > 100) GlobalState.Cache.Clear();

        return true;
    }
}